### ISSUE TITLE
Static queries in templates

### ISSUE THREAD

**Issue by whoisryosuke** [gatsbyjs/gatsby#8395]
## Relevant information

TLDR: Receive error when using templates for generating pages.

Gatsby-node pumps out pages using a template (`/templates/blog-post.js`). When running development server, I receive the following error informing me queries are only executed on **pages**, not **templates**:

```
Exported queries are only executed for Page components. Instead of an exported
query, either co-locate a GraphQL fragment and compose that fragment into the
query (or other fragment) of the top-level page that renders this component, or
use a <StaticQuery> in this component. For more info on fragments and
composition, see http://graphql.org/learn/queries/#fragments and for more
information on <StaticQuery>, see https://next.gatsbyjs.org/docs/static-query
```

How would I go about creating a page component as a template? Is it the same process when using Gatsby's `createPage()` method - just swapping the template file with a page file?

[If you follow the v2 documentation getting started tutorial](https://www.gatsbyjs.org/tutorial/part-seven/), it informs you to make a template -- but if you do this, you'd receive this error message. 

Should the documentation be updated? Is there an example project with the proper methodology?

### Environment (if relevant)

```
  System:
    OS: macOS High Sierra 10.13.4
    CPU: x64 Intel(R) Core(TM) i5-2400S CPU @ 2.50GHz
    Shell: 3.2.57 - /bin/bash
  Binaries:
    Node: 9.8.0 - /usr/local/bin/node
    npm: 6.4.0 - /usr/local/bin/npm
  Browsers:
    Firefox: 61.0.1
  npmPackages:
    gatsby: next => 2.0.0-rc.13
    gatsby-image: next => 2.0.0-rc.1
    gatsby-link: next => 2.0.0-rc.2
    gatsby-mdx: ^0.1.3 => 0.1.3
    gatsby-paginate: 1.0.16 => 1.0.16
    gatsby-plugin-feed: next => 2.0.0-rc.2
    gatsby-plugin-google-analytics: next => 2.0.0-rc.1
    gatsby-plugin-manifest: next => 2.0.2-rc.1
    gatsby-plugin-offline: next => 2.0.0-rc.4
    gatsby-plugin-react-helmet: next => 3.0.0-rc.1
    gatsby-plugin-sharp: next => 2.0.0-rc.3
    gatsby-plugin-sitemap: next => 2.0.0-rc.1
    gatsby-plugin-twitter: next => 2.0.0-rc.2
    gatsby-remark-copy-linked-files: next => 2.0.0-rc.1
    gatsby-remark-images: next => 2.0.1-rc.1
    gatsby-remark-prismjs: next => 3.0.0-rc.2
    gatsby-remark-smartypants: next => 2.0.0-rc.1
    gatsby-source-filesystem: next => 2.0.1-rc.1
    gatsby-transformer-json: next => 2.1.1-rc.2
    gatsby-transformer-react-docgen: next => 2.1.1-rc.5
    gatsby-transformer-remark: next => 2.1.1-rc.1
    gatsby-transformer-sharp: next => 2.1.1-rc.2
  npmGlobalPackages:
    gatsby-cli: 1.1.58
```

### File contents (if changed)

`gatsby-config.js`: N/A 
`package.json`: N/A 
`gatsby-node.js`: 

```
        result.data.allMdx.edges.forEach(async ({ node }) => {
          createPage({
            path: `/${node.frontmatter.menu.toLowerCase()}/${node.parent.name.toLowerCase()}`,
            component: componentWithMDXScope(
              // Should I use a page here instead? Or pass a fragment from here to a page...?
              path.resolve('./src/templates/posts.js'),
              node.code.scope
            ),
            context: {
              id: node.id,
              name: node.frontmatter.name,
            },
          })
        })
```
`gatsby-browser.js`: N/A 
`gatsby-ssr.js`: N/A

**Comment by kakadiadarpan** [comment]
@whoisryosuke can you post a sample template file where you are trying to achieve this? It would be great if you can you provide a reproduction repository for this, that would make it much easier to help you out.

**Comment by CanRau** [comment]
Hm not exactly sure and it's quite late already so might not grasp the whole picture
All `.js` files in src/pages exporting a react component should build without having to add `gatsby-node.js` or extra templates automatically
You only need to manually create pages using `gatsby-node.js` etc.
I think your `gatsby-node.js` and templates are making problems because you don't seem to have `.mdx` files, or are they actually stored outside of the repo? (That's what I do myself)
I'm a little bit confused by your `gatsby-config.js` gatsby-source-filesystem path.. So if you don't have any `.mdx` files yet try to add at least one
Please forgive me if anything is unclear just let me know

**Comment by kakadiadarpan** [comment]
I'm not sure what's the correct way of doing this is. Let's ask @ChristopherBiscardi.

Any suggestions?

**Comment by whoisryosuke** [comment]
Just tested the latest version of [gatsby-starter-blog](https://github.com/gatsbyjs/gatsby-starter-blog) and didn't get the error about using GraphQL in templates. 

Can confirm it does seem like it is dependency-based issue, and I could see it being `gatsby-mdx`.

**Comment by kakadiadarpan** [comment]
@whoisryosuke could you open an issue in [`gatsby-mdx`](https://github.com/ChristopherBiscardi/gatsby-mdx/issues)?

**Comment by ChristopherBiscardi** [comment]
Would be happy to look into it from the gatsby-mdx side. Running through the tutorial might have subtle differences if you replace the remark transformer they use with gatsby-mdx. I'll take a look at trying to replicate this by following the tutorial (tracked in https://github.com/ChristopherBiscardi/gatsby-mdx/issues/155)

For some simple context, the component passed to `componentWithMDXScope` should be the same as the component passed to `createPage`. There is, as far as I know, no substantial difference between having that component in templates vs anywhere else for a 2.0 site. It's hard to say more without the contents of the relevant file, the full error stack output, etc. 

I took a look at the documentation starter and it seems that it has a few issues unrelated to gatsby-mdx so I couldn't replicate this issue:

* filesystem source for components is pointed to a non-existent directory (remove the `..` and replace with `./`
* there's no filesystem source which would pick up mdx files, the only one I can see is the components source
* there's no mdx files so even if a source was set up there would be no graphql results

**Comment by kakadiadarpan** [comment]
There hasn't been any activity on this issue recently. Due to the high number of incoming GitHub issues, we have to clean some of the old issues as many of them have already been resolved with the latest updates.

Please make sure to update to the latest Gatsby version and check if that solves the issue. Let us know if that works for you by adding a comment 👍
